pkg/controllers/polling/controller.go

```diff
@@ -0,0 +1,203 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package polling
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/prometheus/client_golang/prometheus"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"knative.dev/pkg/logging"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	crmetrics "sigs.k8s.io/controller-runtime/pkg/metrics"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/aws/karpenter/pkg/controllers"
+	"github.com/aws/karpenter/pkg/metrics"
+)
+
+// Immediate isn't exactly immediate for a reconcile result. But it should be passed to the RequeueAfter if you want
+// effectively immediate re-reconciliation. This can't be 0 because otherwise controller-runtime won't treat it as a
+// valid RequeueAfter value
+const Immediate = time.Nanosecond
+
+type ControllerInterface interface {
```

Comment:
Worth discussing what it would look like to natively use controller runtime + RequeueAfter. This is nontrivial complexity that we'll need to maintain.

Reply:
Let's discuss. I think the concept of having a single reconciliation loop run that maintains a singleton poll isn't really something that native controller runtime is built for. I think this abstraction layer helps ease the burden of handling the boilerplate of maintaining singleton reconciles while also giving a triggering mechanism for having shorter requeues if needed

Reply:
Agreed for notification, but infrastructure is tied to AWSNodeTemplate and could benefit from the watch.